controllers/vpcresources/cninode_controller.go

```diff
@@ -0,0 +1,128 @@
+// CNINode is deleted when the parent node object is delete ie when the EC2 instance is terminated
+type nodeTerminationFinalizer struct {
+	EC2Wrapper ec2API.EC2Wrapper
+}
+
+var _ finalizer.Finalizer = (*nodeTerminationFinalizer)(nil)
+
+//+kubebuilder:rbac:groups=vpcresources.k8s.aws,resources=cninodes,verbs=get;list;watch;create;update;patch;
+
+// Reconcile handles CNINode create/update/delete events
+// Reconciler will add the finalizer and cluster name tag if it does not exist and finalize on CNINode on deletion to clean up leaked resource on node
+func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cniNode := &v1alpha1.CNINode{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("CNINode is deleted, will be re-created if the node exists", "CNINode", req.NamespacedName)
+		}
+		// Ignore not found error; CNINode will be re-created on node reconcile if the node is not terminated
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add cluster name tag if it does not exist
+	clusterNameTagKey := fmt.Sprintf(config.ClusterNameTagKeyFormat, r.ClusterName)
+	val, ok := cniNode.Spec.Tags[clusterNameTagKey]
+	if !ok || val != config.ClusterNameTagValue {
+		newCNINode := cniNode.DeepCopy()
+		if len(newCNINode.Spec.Tags) != 0 {
+			newCNINode.Spec.Tags[clusterNameTagKey] = config.ClusterNameTagValue
+		} else {
+			newCNINode.Spec.Tags = map[string]string{
+				clusterNameTagKey: config.ClusterNameTagValue,
+			}
+		}
+
+		if err := r.Client.Patch(ctx, newCNINode, client.MergeFrom(cniNode)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to patch CNINode %v", err)
+		}
+	}
+
+	// Finalize will add the finalizer if it does not exist and object is not being delete, and will remove the finalizer if it exists
+	// and object is being deleted. Also sets result.Updated to true so we need to update the object
+	finalizationRes, err := r.finalizers.Finalize(ctx, cniNode)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to finalize CNINode: %v", err)
+	}
+	if finalizationRes.Updated {
+		if err = r.Client.Update(ctx, cniNode); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update CNINode based on finalization result: %w", err)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CNINodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.finalizers = finalizer.NewFinalizers()
+	err := r.finalizers.Register(config.NodeTerminationFinalizer, &nodeTerminationFinalizer{
+		EC2Wrapper: r.EC2Wrapper,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to register the node termination finalizer on CNINode: %v", err)
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.CNINode{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
```

Comment:
Any concern on performance if we only use 1?

Reply:
Updated to use the same number as max concurrent node reconciles